### ISSUE TITLE
armresourcegraph: Inconsistent casing causes resources not to be found from graph API

### ISSUE THREAD

**Issue by TBeijen** [Azure/azure-sdk-for-go#23446]
Inconsistent casing, combined with ids being case-sensitive, causes resources not to be found from graph API.

**Note:** This seems to originate from the graph api itself, since https://portal.azure.com/#view/HubsExtension/ArgQueryBlade shows the same results.

I have a list of resource ids that originate from `armauthorization.RoleAssignmentsClient`, which I take from the responses `Properties.Scope`. I then use armresourcegraph to load the various types of resources to obtain tags.

This works in most cases, however, not for redis. The casing of the ID returned by armauthorization differs from what is expected by the graph API.

```
# ID as returned by armauthorization SDK.
# Does not return result in graph
Resources | where id == '/subscriptions/<my-sub>/resourceGroups/<my-rg>/providers/Microsoft.Cache/redis/<my-redis>' | project id, name, type, resourceGroup, tags

# Fixing casing to the exact type returned when fetching from cli (az resource show --ids "<the-id>")
* Note: redis -> Redis
Resources | where id == '/subscriptions/<my-sub>/resourceGroups/<my-rg>/providers/Microsoft.Cache/Redis/<my-redis>' | project id, name, type, resourceGroup, tags
```
I run use these queries via:

```go
// *armresourcegraph.Client
ResourceGraphClient.Resources(context.Background(), armresourcegraph.QueryRequest{
			Query: to.Ptr(q),
		}, nil)
```

Some inconsistencies can be observed throughout the Azure landscape that make the graph API hard to navigate.

* *Inconsistent strictness of id casing*: When using cli `az resource show --ids "<id-with-lowercase-redis">` this gives the result. The graph API does not. The `az resource` cli command uses dedicated per-resource APIs under the hood. Nevertheless confusing

* *Inconsistent casing of types in ids*: In id `Microsoft.Cache/redis` is incorrect, however `Microsoft.KeyVault/vaults` _is_ correct

* *Inconsistent response casing*: The graph response `type` values are casted to lowercase, unlike the JSON response of `az resource show`. Querying the graph API by type _requires_ this lower case. So querying by id and by type both use strict casing, but require a differently cased value.

I can imagine strict casing is important for all naming entered by users (rg names, resource names, tags, etc.). However the type attribute has a limited and well-defined set of values. Not checking for exact casing would make the armresources SDK and the graph API a lot easier to navigate.

To illustrate, I now have this code fragment to work around this issue:
```
fixedResourceID := strings.Replace(resourceID, "Microsoft.Cache/redis", "Microsoft.Cache/Redis", -1)
```

**Comment by tadelesh** [comment]
@chiragg4u is it by design that resource type is case sensitive? @JeffreyRichter could you also help to take a look? is their any standard for azure resource id/resource type of the casing?

**Comment by JeffreyRichter** [comment]
According to the industry standard RFC for URLs (https://datatracker.ietf.org/doc/html/rfc3986#section-6.2.2.1), URLs are supposed to be case-sensitive except for the scheme and host name). 

Azure services should honor this, but (unfortunately) some do not, and this causes these kinds of problems. And because some engineers are not aware of the RFC, they *think* URLs are case-insensitive and so one engineer might return "redis" while another engineer writes code requiring "Redis".

So, case-sensitive comparisons are the correct thing. The team needs to agree on "redis" vs "Redis" and then stick to it consistently. This may require code changes. As a "quick-fix", a service could treat "redis" and "Redis" as the same (and maybe "REDIS" too) but this can lead to more problems in the future.

**Comment by TBeijen** [comment]
Agreed that in urls, ultimately differently cased ids are different things.

I stumbled upon another one: assignments sometimes referencing scope of type subscription in all uppercase: `"/subscriptions/1000AAAA-BBBB-CCCC-DDDD-123456789012"`

I can imagine this will be quite hard to fix, given the enormous scope of all the APIs, different teams, etc. 

Probably some sort of normalize service/library that can be used to validate/normalize resource ids on input, compare output with normalized (and optionally fix).... (Just thinking out loud, it's easier said than done and has potential to cause all kinds of little new bugs in consumer's code)
